### PR TITLE
Add unit test coverage for P2P PeerPool - Closes #4229

### LINKED ISSUE #4229: Add unit test coverage for P2P peer pool file
### Expected behavior

Every file from `lisk-p2p` library is well covered with unit test. The minimum must be 80% at least to achieve green reports.  

### Actual behavior

`peer_pool.ts` counts with only  76.32 %.

### Steps to reproduce

`npm t`

### Which version(s) does this affect? (Environment, OS, etc...)

`+2.3`

### PR DESCRIPTION
**Pull request by mitsujutsu**
### What was the problem?

Unit tests were missing for `PeerPool` in `lisk-p2p` library.

### How did I solve it?

Added missing unit tests.

### How to manually test it?

npm run test

### Review checklist

- [ ] The PR resolves #4229 
- [ ] All new code is covered with unit tests
- [ ] Relevant integration / functional tests are added
- [ ] Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
- [ ] Documentation has been added/updated

### CODE REVIEW

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 1]
```diff
@@ -71,14 +82,523 @@ describe('peerPool', () => {
...
 		},
+	};
+	let peerPool: PeerPool;
+	let peerInfo: P2PDiscoveredPeerInfo;
+	let nodeInfo: P2PNodeInfo;
+	let peerId: string;
+	let peerObject: any;
+	let messagePacket: any;
+	let requestPacket: any;
+	beforeEach(async () => {
```
Comment (ishantiw) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: new line before `beforeEach`

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 2]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+			on: sandbox.stub(),
+			off: sandbox.stub(),
+			emit: sandbox.stub(),
+			destroy: sandbox.stub(),
+		} as any;
+		peerPool.emit = sandbox.stub().resolves();
+	});
+
+	afterEach(async () => {
+		sandbox.restore();
```
Comment (shuse2) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: isn't this call from _global_hook?

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 3]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+				requestPacket,
+			});
+		});
+
+		it('should throw error if no peers selected', async () => {
+			sandbox.stub(peerPool as any, '_peerSelectForRequest').returns([]);
+			try {
+				await peerPool.request(requestPacket);
+			} catch (err) {
+				expect(err).to.be.instanceOf(RequestFailError);
```
Comment (shuse2) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: this could be false positive if the request doesn't throw.
It's should have
```
let err;
try {
...
} catch (error) {
  err = error;
}
expect(err).to.be.instanceOf(RequestFailError);
```

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 4]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+				expect(peerPool.emit).to.be.calledOnce;
+			}
+		});
+	});
+
+	describe('#requestFromPeer', () => {
+		it('should throw error if no peers in peerPool', async () => {
+			(peerPool as any)._peerMap = new Map();
+			try {
+				await peerPool.requestFromPeer(requestPacket, peerId);
```
Comment (shuse2) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: same as above

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 5]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+
+			expect(peerObject.request).to.be.calledWithExactly(requestPacket);
+		});
+	});
+
+	describe('#sendToPeer', () => {
+		it('should throw error if no peers in peerPool', async () => {
+			(peerPool as any)._peerMap = new Map();
+			try {
+				await peerPool.sendToPeer(messagePacket, peerId);
```
Comment (shuse2) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: same as above

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 6]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+			sandbox.stub(peerPool as any, '_evictPeer');
+			peerPool.addInboundPeer(peerInfo, peerObject as any);
+
+			expect((peerPool as any)._evictPeer).to.be.calledWithExactly(InboundPeer);
+		});
+
+		it('should throw error if peer already exists in peerpool', async () => {
+			(peerPool as any)._peerMap = new Map([[peerId, peerObject]]);
+			try {
+				peerPool.addInboundPeer(peerInfo, peerObject as any);
```
Comment (shuse2) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: same as above

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 7]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+			peerPool.addOutboundPeer(peerId, peerObject as any);
+
+			expect(getPeerStub).to.be.calledWithExactly(peerId);
+		});
+
+		it('should return peer if peer already exists in peerpool', async () => {
+			(peerPool as any)._peerMap = new Map([[peerId, peerObject]]);
+
+			try {
+				peerPool.addOutboundPeer(peerId, peerObject as any);
```
Comment (shuse2) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: same as above

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 8]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+			on: sandbox.stub(),
+			off: sandbox.stub(),
+			emit: sandbox.stub(),
+			destroy: sandbox.stub(),
+		} as any;
+		peerPool.emit = sandbox.stub().resolves();
+	});
+
+	afterEach(async () => {
+		clock.restore();
```
Comment (diego-G) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: We shouldn't have to do this.

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 9]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
 	});
 
 	describe('#constructor', () => {
-		it('should be an object and instance of PeerPool', () => {
-			return expect(peerPool)
+		it('should be an object and instance of PeerPool', async () => {
+			expect(peerPool)
 				.to.be.an('object')
```
Comment (diego-G) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: Redundant.

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 10]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+				nodeInfo: peerPool.nodeInfo,
+				peerLimit: 1,
+				requestPacket,
+			});
+		});
+
+		it('should throw error if no peers selected', async () => {
+			sandbox.stub(peerPool as any, '_peerSelectForRequest').returns([]);
+			let err;
+			try {
```
Comment (diego-G) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: Prefer to use `.to.be.rejectedWith` instead of using `try/catch`.

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 11]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+
+			expect(sendToPeer).to.be.calledOnceWithExactly(
+				messagePacket,
+				constructPeerIdFromPeerInfo(peerInfo),
+			);
+		});
+
+		it(`should emit event if sendToPeer fails`, async () => {
+			sendToPeer.throws();
+			try {
```
Comment (diego-G) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: Why do we use `try/catch`?

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 12]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+		it('should throw error if no peers in peerPool', async () => {
+			(peerPool as any)._peerMap = new Map();
+			let err;
+			try {
+				await peerPool.requestFromPeer(requestPacket, peerId);
+			} catch (error) {
+				err = error;
+			}
+
+			expect(err).to.exist;
```
Comment (diego-G) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: Same as above.

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 13]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+
+			expect(peerObject.request).to.be.calledWithExactly(requestPacket);
+		});
+	});
+
+	describe('#sendToPeer', () => {
+		it('should throw error if no peers in peerPool', async () => {
+			(peerPool as any)._peerMap = new Map();
+			let err;
+			try {
```
Comment (diego-G) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: Same as above.

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 14]
```diff
@@ -71,14 +81,539 @@ describe('peerPool', () => {
...
+			sandbox.stub(peerPool as any, '_evictPeer');
+			peerPool.addInboundPeer(peerInfo, peerObject as any);
+
+			expect((peerPool as any)._evictPeer).to.be.calledWithExactly(InboundPeer);
+		});
+
+		it('should throw error if peer already exists in peerpool', async () => {
+			(peerPool as any)._peerMap = new Map([[peerId, peerObject]]);
+			let err;
+			try {
```
Comment (diego-G) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: Please, try to remove every `try/catch` as I've proposed above.

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 15]
```diff
@@ -386,60 +921,6 @@ describe('peerPool', () => {
...
-			expect(peerPool['_peerSelectForSend']).to.be.calledWith({
-				peers: [],
-				nodeInfo: peerPool.nodeInfo,
-				peerLimit: DEFAULT_SEND_PEER_LIMIT,
-				messagePacket: { event: 'foo', data: 123 },
-			});
-		});
-	});
-
 	describe('#triggerNewConnections', () => {
```
Comment (diego-G) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: Could you order these describe sections as they exist in `peep_pool.ts` file? Please, add skeletons if needed for those methods not addressed in this issue.

**Comment by mitsujutsu** [comment]
> Few more comments:
> 
>     * There are several uncovered lines as it can be seen in the coverage report `(...79,502,589,675)`.
> 
>     * There is a skipped test that should be re-enabled `describe.skip('#getConnectedPeers'`
> 
>     * When running the test file standalone, the coverage is quite low
>       `peer_pool.ts      |    74.58 |    51.35 |    53.33 |    75.11 |... 89,690,691,692`
>       and the following error happens:
>       `TS_NODE_PROJECT=./test/tsconfig.json npx nyc npx mocha test/unit/peer_pool.ts`
> 
> 
> ```
>       at checkWrappedMethod (/Users/diego-g/git/lisk-sdk/node_modules/sinon/lib/sinon/util/core/wrap-method.js:39:21)
>       at wrapMethod (/Users/diego-g/git/lisk-sdk/node_modules/sinon/lib/sinon/util/core/wrap-method.js:87:13)
>       at stub (/Users/diego-g/git/lisk-sdk/node_modules/sinon/lib/sinon/stub.js:73:44)
>       at Sandbox.stub (/Users/diego-g/git/lisk-sdk/node_modules/sinon/lib/sinon/sandbox.js:291:33)
>       at Context.beforeEach (test/unit/peer_pool.ts:694:6)
> ```

There were some functions for which were already covered, and you had mentioned to do `triggerNewConnections`. Due to these functions which I did not touch, coverage may be low.

File: elements/lisk-p2p/test/unit/peer_pool.ts (hunk 8 above)
Comment (mitsujutsu) [reply elements/lisk-p2p/test/unit/peer_pool.ts]: It is required otherwise the test hangs.

File: elements/lisk-p2p/test/unit/peer_pool.ts (hunk 14 above)
Comment (mitsujutsu) [reply elements/lisk-p2p/test/unit/peer_pool.ts]: In this case it is required as `toBeRejectedWith` and `throws` does not work.

File: elements/lisk-p2p/test/unit/peer_pool.ts (hunk 12 above)
Comment (mitsujutsu) [reply elements/lisk-p2p/test/unit/peer_pool.ts]: `try catch` required as toBeRejectedWith and throws does not work.

File: elements/lisk-p2p/test/unit/peer_pool.ts (hunk 8 above)
Comment (diego-G) [reply elements/lisk-p2p/test/unit/peer_pool.ts]: You're right. But, perhaps the test failure when running this file standalone is related to this. Nevermind otherwise.

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 16]
```diff
@@ -273,14 +272,10 @@ describe('peerPool', () => {
...
 			sandbox.stub(peerPool as any, '_peerSelectForRequest').returns([]);
-			let err;
-			try {
-				await peerPool.request(requestPacket);
-			} catch (error) {
-				err = error;
-			}
 
-			expect(err).to.be.instanceOf(RequestFailError);
+			expect(peerPool.request(requestPacket)).to.be.rejectedWith(
```
Comment (diego-G) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: We need `return` here to avoid unhandled errors.

File: elements/lisk-p2p/test/unit/peer_pool.ts [hunk 17]
```diff
@@ -335,14 +328,10 @@ describe('peerPool', () => {
...
 			(peerPool as any)._peerMap = new Map();
-			let err;
-			try {
-				await peerPool.requestFromPeer(requestPacket, peerId);
-			} catch (error) {
-				err = error;
-			}
 
-			expect(err).to.exist;
+			expect(
```
Comment (diego-G) [inline elements/lisk-p2p/test/unit/peer_pool.ts]: Same as above.

File: elements/lisk-p2p/test/unit/peer_pool.ts (hunk 16 above)
Comment (shuse2) [reply elements/lisk-p2p/test/unit/peer_pool.ts]: Since it's async, i think `return` is not needed?

Comment (diego-G) [reply elements/lisk-p2p/test/unit/peer_pool.ts]: @shuse2 I thought after our discussion, it was clear that we need it. Please, check it out again. If we change the expectation to another kind of error for example, we obtain an `UnhandledPromiseRejectionWarning: AssertionError` with a false positive.

@mitsujutsu check it out for yourself too. I suggest to include as well the message error like this:
```
			expect(peerPool.request(requestPacket)).to.eventually.be.rejectedWith(
				RequestFailError,
				'Request failed due to no peers found in peer selection',
			);
```

File: templates/test/mocha.opts.tmpl [hunk 18]
```diff
@@ -3,4 +3,5 @@
 --require tsconfig-paths/register
 --require source-map-support/register
 --require ./test/_setup.ts
+--file ./test/unit/_global_hooks.ts
```
Comment (shuse2) [inline templates/test/mocha.opts.tmpl]: No need to add this here, and it will break other tests

Comment (mitsujutsu) [reply templates/test/mocha.opts.tmpl]: @diego-G pointed out that a test breaks when run individually with `npx mocha` due to sandbox not being cleaned properly. so in case someone wants to run the test individually like `TS_NODE_PROJECT=./test/tsconfig.json npx nyc npx mocha test/unit/peer_pool.ts` there is a setup missing somewhere. running `npm run test` normally works fine. I think it is a bit of edge case but to satisfy running the test specifically via `TS_NODE_PROJECT=./test/tsconfig.json npx nyc npx mocha test/unit/peer_pool.ts` I will add an extra `sandbox.restore` in the `beforeEach` of the file. If you see this extra restore in the code this is the reason why. @diego-G

Comment (shuse2) [reply templates/test/mocha.opts.tmpl]: but most of the library doesn't have this structure, so it will not pass the other tests.
Also, _global_hook is automatically caught by the mocha runtime

Comment (diego-G) [reply templates/test/mocha.opts.tmpl]: I've added a commit to solve the issue. @shuse2 @mitsujutsu
